File: java/binarysearchtree/src/main/java/com/js/datastructuresalgorithms/bst/JSBinarySearchTree.java
package com.js.datastructuresalgorithms.bst;

public class JSBinarySearchTree<T extends Comparable<T>> {

    //*****************
    // Class Members **
    //*****************
    private Node mRoot;
    private class Node {
        private T mValue;
        private Node mLeft;
        private Node mRight;

        private Node(T value, Node left, Node right) {
            mValue = value;
            mLeft = left;
            mRight = right;
        }

        private T value() { return mValue; }
        private void value(T value) { mValue = value; }

        private Node left() { return mLeft; }
        private void left(Node left) { mLeft = left; }

        private Node right() { return mRight; }
        private void right(Node right) { mRight = right; }
    }

    public void add(T value) {
        if(root() == null) {
            mRoot = new Node(value, null, null);
        } else {
            add(root(), value);
        }
    }

    public void clear() { mRoot = null; }

    public boolean contains(T value) { return findNode(value) != null; }

    public boolean remove(T value) { return remove(findNode(value)); }

    //*******************
    // Private Members **
    //*******************

    private boolean remove(Node node)
    {
        Node parent = findParent(node);

        if (parent != null) {
            if (node.right() != null && node.left() != null) {
                Node min = min(node.right());
                remove(min);
                node.value(min.value());
            } else if (node.right() == null && node.left() == null) {
                deleteChild(parent, node);
            } else {
                moveChild(parent, node);
            }
        }

        return parent != null;
    }

    private void moveChild(Node parent, Node child)
    {
        Node update = child.right() == null ? child.left() : child.right();
        if(equal(parent.left(), child)) {
            parent.left(update);
        } else if(equal(parent.right(), child)) {
            parent.right(update);
        }
    }

    private Node min(Node node) {
        while(node.left() != null) {
            node = node.left();
        }
        return node;
    }

    private void deleteChild(Node parent, Node child) {
        if(equal(parent.left(), child)) {
            parent.left(null);
        } else if(equal(parent.right(), child)) {
            parent.right(null);
        }
    }

    private Node findParent(Node node)
    {
        Node next = root();
        while(!isParent(next, node)) {
            next = whichWay(next, node);
        }
        return next;
    }

    private Node whichWay(Node node, Node child) {
        return compare(node, child) > 0 ? node.left() : node.right();
    }

    private boolean isParent(Node node, Node child)
    {
        return equal(node.left(), child) || equal(node.right(), child);
    }

    private Node findNode(T value) {
        Node next = root();
        while(next != null && !equal(next, value)) {
            next = compare(next, value) > 0 ? next.left() : next.right();
        }
        return next;
    }

    private int compare(Node one, T value) {
        return one.value().compareTo(value);
    }

    private int compare(Node one, Node two) {
        return one.value().compareTo(two.value());
    }

    private boolean equal(Node one, Node two) {
        return one == two || one != null && one.value().compareTo(two.value()) == 0;
    }

    private boolean equal(Node one, T value) {
        return one != null && one.value().compareTo(value) == 0;
    }

    private void add(Node node, T value)
    {
        int compare = node.value().compareTo(value);
        if(compare > 0) {
            if(node.left() == null) {
                node.left(new Node(value, null, null));
            } else {
                add(node.left(), value);
            }
        } else if(compare < 0) {
            if(node.right() == null) {
                node.right(new Node(value, null, null));
            } else {
                add(node.right(), value);
            }
        }
    }

    private Node root() { return mRoot; }
}

File: node/binary-search-tree/src/types/JSBinarySearchTree.js
class Node {
    #mLeft
    #mRight
    #mValue

    constructor(value) {
        this.#mValue = value;
        this.#mLeft = null;
        this.#mRight = null;
    }

    get value() { return this.#mValue; }
    set value(value) { this.#mValue = value; }

    get left() { return this.#mLeft; }
    set left(value) { this.#mLeft = value; }

    get right() { return this.#mRight; }
    set right(value) { this.#mRight = value; }
}

class JSBinarySearchTree {

    #mRoot

    constructor() { this.#mRoot === null; }

    clear = () => this.#mRoot = null;

    add = (item) => {
        if(this.#root() === null) {
            this.#mRoot = new Node(item);
        } else {
            this.#addItem(item);
        }
    }

    contains = (item) => this.#findNode(item) !== null;

    remove = (item) => this.#removeNode(this.#findNode(item));

    preorder = () => this.#doPreorder(this.#root());

    inOrder = () => this.#doInOrder(this.#root());



    //debug = () => this.#root();

    //*******************
    // Private Members **
    //*******************

    #doPreorder = (node) => {
        if (node === null || node === undefined) return;
        console.log(node.value);
        this.#doPreorder(node.left);
        this.#doPreorder(node.right);
    }

    #doInOrder = (node) => {
        if (node === null || node === undefined) return;
        this.#doInOrder(node.left);
        console.log(node.value);
        this.#doInOrder(node.right);
    }

    #removeNode = (node) => {
        if(node !== null) {
            const parent = this.#parent(node);
            if(node.left === null && node.right === null) {
                this.#deleteChild(parent, node);
            } else if(node.left !== null && node.right !== null) {
                const lowestRight = this.#lowestRight(node);
                this.#removeNode(lowestRight);
                node.value = lowestRight.value;
            } else {
                if(node.left !== null) {
                    parent.left === node ? parent.left = node.left : parent.right = node.left;
                } else if(node.right !== null) {
                    parent.left === node ? parent.left = node.right : parent.right = node.right;
                }
            }
        }
    }

    #lowestRight = (node) => {
        node = node.right;
        while(node.left !== null) {
            node = node.left;
        }
        return node;
    }

    #deleteChild = (parent, child) => {
        if(this.#equal(parent.left, child)) {
            parent.left = null;
        } else if(this.#equal(parent.right, child)) {
            parent.right = null;
        }
    }

    #parent = (child) => {
        let node = this.#root();
        while(!this.#isParent(node, child)) {
            node = child.value > node.value ? node.right : node.left;
        }
        return node;
    }

    #isParent = (node, child) => node !== null && (node.left === child || node.right === child);

    #equal = (one, two) => (one === two) || (one !== null && two !== null && one.value === two.value);

    //#equals = (node, value) => node !== null && node.value === value;

    #findNode = (item) => {
        let node = this.#root();
        while(node !== null && node.value !== item) {
            node = item > node.value ? node.right : node.left;
        }
        return node;
    }

    #addItem = (item) => {
        let node = this.#root();
        let child = node;
        while(child !== null) {
            node = child;
            child = item > node.value ? node.right : node.left;
        }
        if(item > node.value) {
            node.right = new Node(item);
        } else if(item < node.value) {
            node.left = new Node(item);
        }
    }

    #root = () => this.#mRoot;
}

module.exports = JSBinarySearchTree;